diff --git a/dist/react-contexify.cjs.development.js b/dist/react-contexify.cjs.development.js
index 8b91bec..cf2b3ea 100644
--- a/dist/react-contexify.cjs.development.js
+++ b/dist/react-contexify.cjs.development.js
@@ -70,7 +70,7 @@ function createEventManager() {
       return this;
     },
     emit: function emit(event, args) {
-      if (process.env.NODE !== 'production') {
+      {
         var currentEv = event;
 
         if (!eventList.has(event) && currentEv !== 0
diff --git a/dist/react-contexify.esm.js b/dist/react-contexify.esm.js
index 804292d..cfb88dc 100644
--- a/dist/react-contexify.esm.js
+++ b/dist/react-contexify.esm.js
@@ -63,7 +63,7 @@ function createEventManager() {
       return this;
     },
     emit: function emit(event, args) {
-      if (process.env.NODE !== 'production') {
+      if (process.env.NODE_ENV !== 'production') {
         var currentEv = event;
 
         if (!eventList.has(event) && currentEv !== 0
diff --git a/src/core/eventManager.ts b/src/core/eventManager.ts
index ace1493..a4e6381 100644
--- a/src/core/eventManager.ts
+++ b/src/core/eventManager.ts
@@ -25,7 +25,7 @@ function createEventManager<E = EventType>(): EventManager<E> {
       return this;
     },
     emit<T = any>(event: E, args: T) {
-      if (process.env.NODE !== 'production') {
+      if (process.env.NODE_ENV !== 'production') {
         const currentEv = (event as unknown) as number;
 
         if (!eventList.has(event) && currentEv !== EVENT.HIDE_ALL) {
